src/fixed_decimals_utils.jl

```diff
@@ -127,284 +131,304 @@ _typemaxbytes(::Type{Int128}, i, is_neg) = @inbounds NTuple{39,UInt8}((0x31, 0x3
-    Parsers.invalid(code) && return (T(0), code, pos)
-
-    # if we got here, we can safely ignore all bytes that == groupmark, these also must appear before decimal_position and exp_position
-    groupmark = something(options.groupmark, 0xff)
-
-    last_byte_to_parse = max(pos, exp_position == 0 ? field_end - !Parsers.eof(code) - Parsers.quoted(code) : exp_position - 1)
-    if exp_position != 0
-        exp_result = Parsers.xparse(Int, buf, exp_position+1, field_end - Parsers.quoted(code), options)
-        exp_val = exp_result.val
-        !Parsers.ok(exp_result.code) && return (T(0), code |= exp_result.code, field_end)
+        Parsers.incr!(source)
+        Parsers.eof(source, pos, len) && @goto eof_invalid # Input is something like "[-+]<EOF>"
+        b = Parsers.peekbyte(source, pos)
     else
-        exp_val = 0
+        Parsers.eof(source, pos, len) && @goto eof_invalid # Input is something like "<EOF>" (empty)
     end
 
-    parse_through_decimal = decimal_position != 0
-    has_digits_past_decimal = parse_through_decimal
-    number_of_fractional_digits = 0
-    if has_digits_past_decimal
-        # Remove trailing zeros after decimal
-        while buf[last_byte_to_parse] == UInt8('0')
-            last_byte_to_parse -= 1
+    last_byte_to_parse = 0
+    first_byte_to_parse = Int(pos + (b == options.decimal)) # will be adjusted later if we have leading zeros
+
+    # The digits before the decimal point / exponent
+    prev_b = groupmark + 0x01
+    if UInt8('0') <= b <= UInt8('9')
+        found_digit = true
+        prev_b = b
+        if b != UInt8('0')
+            num_digits += 1
+            found_nonzero = true
+            last_nonzero_digit_pos = pos
         end
-        if last_byte_to_parse == decimal_position
-            parse_through_decimal = false
-            has_digits_past_decimal = false
-            last_byte_to_parse -= 1
+        pos += 1
+        Parsers.incr!(source)
+        Parsers.eof(source, pos, len) && (code |= Parsers.EOF; @goto done_scanning)
+        while true
+            b = Parsers.peekbyte(source, pos)
+            if b == UInt8('0')
+                # Skip leading zeros
+                found_nonzero && (num_digits += 1)
+                pos += 1
+                Parsers.incr!(source)
+                Parsers.eof(source, pos, len) && (code |= Parsers.EOF; @goto done_scanning) # Input is something like "10<EOF>"
+            elseif UInt8('1') <= b <= UInt8('9')
+                !found_nonzero && (first_byte_to_parse = pos)
+                found_nonzero = true
+                last_nonzero_digit_pos = pos
+                pos += 1
+                Parsers.incr!(source)
+                num_digits += 1
+                Parsers.eof(source, pos, len) && (code |= Parsers.EOF; @goto done_scanning) # Input is something like "11<EOF>"
+            elseif has_groupmark && b == groupmark
+                # More than one groupmark in a row is invalid
```

Comment:
was confused by "in a row" for a second there (thinking about e.g. a row in a CSV file)
```suggestion
                # More than one groupmark consecutively is invalid
```